Mathlib/Algebra/Group/Prod.lean

```diff
@@ -574,38 +574,89 @@ def prodUnique [Unique N] : M × N ≃* M :=
+  monoid to the additive units of its product on the left with another."]
+def inr : Nˣ →* (M × N)ˣ := Units.map (MonoidHom.inr M N)
+
+@[to_additive (attr := simp)]
+theorem fst_comp_inl  : fst.comp (inl (N := N)) = MonoidHom.id Mˣ := rfl
+@[to_additive (attr := simp)]
+theorem fst_inl (u : Mˣ) : fst (inl (N := N) u) = u := rfl
+@[to_additive (attr := simp)]
+theorem snd_comp_inl : snd.comp inl = (1 : Mˣ →* Nˣ) := rfl
+@[to_additive (attr := simp)]
+theorem snd_inl (u : Mˣ) : snd (inl (N := N) u) = 1 := rfl
+@[to_additive (attr := simp)]
+theorem fst_comp_inr : fst.comp inr = (1 : Nˣ →* Mˣ) := rfl
+@[to_additive (attr := simp)]
+theorem fst_inr (u : Nˣ) : fst (inr (M := M) u) = 1 := rfl
+@[to_additive (attr := simp)]
+theorem snd_comp_inr : snd.comp (inr (M := M)) = MonoidHom.id Nˣ := rfl
+@[to_additive (attr := simp)]
+theorem snd_inr (u : Nˣ) : snd (inr (M := M) u) = u := rfl
+
+/-- A map from the product of the units of two monoids to the units of their product. -/
+@[to_additive prod "A map from the product of the additive units of two
+    additive monoids to the additive units of their product."]
+def prod : Mˣ × Nˣ →* (M × N)ˣ :=
+  MonoidHom.toHomUnits (MonoidHom.prodMap (Units.coeHom M) (Units.coeHom N))
+
+@[to_additive (attr := simp) val_prod_apply]
+theorem val_prod_apply (g : Mˣ × Nˣ) : (prod g).val = (g.1.val, g.2.val) := rfl
+
+@[to_additive (attr := simp) val_inv_prod_apply]
+theorem val_inv_prod_apply (g : Mˣ × Nˣ) : (prod g)⁻¹.val = (g.1⁻¹.val, g.2⁻¹.val) := rfl
+
+@[to_additive (attr := simp) fst_prod]
+theorem fst_prod (g : Mˣ × Nˣ) : (prod g).fst = g.1 := rfl
+
+@[to_additive (attr := simp) snd_prod]
+theorem snd_prod (g : Mˣ × Nˣ) : (prod g).snd = g.2 := rfl
+
+@[to_additive (attr := simp) prod_fst_snd]
+theorem prod_fst_snd (g : (M × N)ˣ) : prod (fst g, snd g) = g := rfl
+
+@[to_additive]
+lemma _root_.Prod.isUnit_iff {x : M × N} : IsUnit x ↔ IsUnit x.1 ∧ IsUnit x.2 where
+  mp h := ⟨h.unit.fst.isUnit, h.unit.snd.isUnit⟩
+  mpr h := (prod (h.1.unit, h.2.unit)).isUnit
+
 /-- The monoid equivalence between units of a product of two monoids, and the product of the
     units of each monoid. -/
 @[to_additive prodAddUnits
       "The additive monoid equivalence between additive units of a product
       of two additive monoids, and the product of the additive units of each additive monoid."]
 def prodUnits : (M × N)ˣ ≃* Mˣ × Nˣ where
-  toFun := (Units.map (MonoidHom.fst M N)).prod (Units.map (MonoidHom.snd M N))
-  invFun u := ⟨(u.1, u.2), (↑u.1⁻¹, ↑u.2⁻¹), by simp, by simp⟩
-  left_inv u := by
-    simp only [MonoidHom.prod_apply, Units.coe_map, MonoidHom.coe_fst, MonoidHom.coe_snd,
-      Prod.mk.eta, Units.coe_map_inv, Units.mk_val]
-  right_inv := fun ⟨u₁, u₂⟩ => by
-    simp only [Units.map, MonoidHom.coe_fst, Units.inv_eq_val_inv,
-      MonoidHom.coe_snd, MonoidHom.prod_apply, Prod.mk.injEq]
-    exact ⟨rfl, rfl⟩
+  toFun := fst.prod snd
+  invFun := prod
```

Comment:
Can you please add the simp lemma that `prodUnits.symm.toMonoidHom = Units.prod`?

Which also helps me see, you changed the namespace for `prodUnits`.

Reply:
Oh, yes. Should I put it back in `MulEquiv`?

Reply:
We seem to often have Units-related Equivs in `Units`, e.g. `Units.mapEquiv`.

Reply:
Makes sense, just noting the API surface change (perhaps a deprecation?)

Reply:
It's also why you had a build failure:
```
error: /home/lean/actions-runner/_work/mathlib4/mathlib4/Mathlib/Data/Nat/Totient.lean:131:26: unknown constant 'MulEquiv.prodUnits'
```